File: 1222-remove-covered-intervals/remove-covered-intervals.java
import java.util.Arrays;

class Solution {
    public int removeCoveredIntervals(int[][] intervals) {
        // WHAT: Sort intervals: ascending by start point; descending by end point if starts match.
        // WHY: If starts match, the wider interval comes first, covering any subsequent narrower intervals.
        Arrays.sort(intervals, (a, b) -> {
            if (a[0] != b[0]) {
                return Integer.compare(a[0], b[0]);
            } else {
                return Integer.compare(b[1], a[1]);
            }
        });

        int remainingCount = 0;
        int maxEnd = 0;

        // WHAT: Traverse each interval and compare its end point with the maximum end point seen so far.
        for (int[] interval : intervals) {
            int currentEnd = interval[1];

            // WHAT: If currentEnd fits inside maxEnd, it means this interval is completely covered.
            // WHY: Due to sorting, we already know its start point is >= the start point of the interval that formed maxEnd.
            if (currentEnd <= maxEnd) {
                continue; // Covered, drop it
            } else {
                // Not covered, so it stays as a unique remaining interval
                remainingCount++;
                maxEnd = currentEnd; // Update the boundary tracker
            }
        }

        return remainingCount;
    }
}

File: 1222-remove-covered-intervals/remove-covered-intervals.py
class Solution:
    def removeCoveredIntervals(self, intervals: list[list[int]]) -> int:
        # WHAT: Sort intervals: ascending by start point (x[0]), descending by end point (-x[1]).
        # WHY: This ensures that for identical start points, the longest interval is processed first to catch and cover shorter ones.
        intervals.sort(key=lambda x: (x[0], -x[1]))

        remaining_count = 0
        max_end = 0

        # WHAT: Linearly scan the sorted list of intervals.
        for _, end in intervals:
            # WHAT: If the current end point is less than or equal to max_end, it is covered.
            # WHY: The sorting order ensures its start point is validly contained within a previously evaluated larger interval.
            if end <= max_end:
                continue  # Covered interval, skip counting it
            else:
                # This interval extends past the previous ones, making it unique and valid
                remaining_count += 1
                max_end = end  # Stretch our coverage boundary forward

        return remaining_count